### DISCUSSION TITLE
How to read file from disk?

### DISCUSSION THREAD

**Discussion by nyssance** [tauri-apps/tauri#10915] [Q&A]
I am from Electron to Tauri 2.0

I use svelte, and if when it's run at localhost, work with default url(https://localhost:1420/data/list.json). If run from app(Electron/Tauri)，I register a protocol, catch the scheme and change uri to local path(mac ~/Datacenter/list.json or windows E:\\Datacenter\list.json).

this is my lib.rs, I can get the uri, how to get file and return the json? all documents is js, no rust version.

```rs
#[tauri::command]
fn greet(name: &str) -> String {
    format!("Hello, {}! You've been greeted from Rust!", name)
}

#[cfg_attr(mobile, tauri::mobile_entry_point)]
pub fn run() {
    tauri::Builder::default()
        .plugin(tauri_plugin_deep_link::init())
        .plugin(tauri_plugin_fs::init())
        .register_asynchronous_uri_scheme_protocol("app", |_app, request, responder| {
            let uri = request.uri();  // like app://datacenter/aaa/bbb.json
            println!("{}", uri);
            // TODO, change uri to ~/Datacenter/aaa/bbb.json and parse and return
        })
        .plugin(tauri_plugin_shell::init())
        .invoke_handler(tauri::generate_handler![greet])
        .run(tauri::generate_context!())
        .expect("error while running tauri application");
}
```

svelte code like this
```js
export const getFileUrl = (file, url) => {
    // https://github.com/electron/electron/issues/2288#issuecomment-1317089683
    const isElectron = navigator.userAgent.indexOf('Electron') >= 0
    const isTauri = '__TAURI_INTERNALS__' in window
    const isApp = isElectron || isTauri
    return `${isApp ? 'app://datacenter' : url.origin}${file}`
}
```

```js
export async function load({ fetch, url }) {
    let products = undefined
    try {
        const file = '/data/list.json'
        const response = await fetch(getFileUrl(file, url))
        products = await response.json()
    } catch (err) { }
    return { products }
}
```
How to fetch the json data when work in Tauri 2.0?

**Comment by FabianLars** [comment]
Did you know about the built-in `asset://` / `http://asset.localhost/` protocol <https://v2.tauri.app/reference/javascript/api/namespacecore/#convertfilesrc>? If you don't need to modify the files that one may already work for you.

Otherwise the docs seem to show exactly what you're asking for, no? https://docs.rs/tauri/2.0.0-beta/tauri/struct.Builder.html#method.register_uri_scheme_protocol
